# Capture last-token residual

**Previous:** [00_OVERVIEW](../00_OVERVIEW.ipynb)  
**Home:** [../../00_START_HERE.ipynb](../../00_START_HERE.ipynb)  
**Next:** [02_positions](02_positions.ipynb)

**Kernel:** `CXR local Qwen (faiss_gpu1)`  
**Status:** executable (calls `scripts/` — same as CLI)

---

## 1. Why am I learning this?

Executable Observe lesson — same as `look.py --mode capture`.

## 2. Mental model

Read activations / logits. Do not write the forward.

## 3. Clinical question

What does this mode show on the built-in oncology prompts?

## 4. Prediction

Non-trivial numbers / strings; see CLI output style.


## 5. Minimal Python — setup + load (run once)


In [2]:
# Shared bootstrap — run this first in every executable notebook
import sys
from pathlib import Path

# notebooks/_lib regardless of how deep this notebook sits
_here = Path.cwd().resolve()
for _p in [_here, *_here.parents]:
    if (_p / "_lib" / "cxr_boot.py").is_file():
        sys.path.insert(0, str(_p / "_lib"))
        break
    if (_p / "notebooks" / "_lib" / "cxr_boot.py").is_file():
        sys.path.insert(0, str(_p / "notebooks" / "_lib"))
        break
else:
    raise RuntimeError("Cannot find notebooks/_lib/cxr_boot.py — open Jupyter with notebooks/ as root")

import cxr_boot
ctx = cxr_boot.setup(layer=20, max_new=24, load_model=False)
model, tok = ctx["model"], ctx["tok"]
NOTE, LAYER, MAX_NEW = ctx["NOTE"], ctx["LAYER"], ctx["MAX_NEW"]
PROMPT_A, PROMPT_B, PROMPT_TEST = ctx["PROMPT_A"], ctx["PROMPT_B"], ctx["PROMPT_TEST"]
look, intervene, process = ctx["look"], ctx["intervene"], ctx["process"]
print("NOTE:", NOTE)
print("LAYER:", LAYER, "ready")


Qwen already loaded on :8270 pid=762992 (Qwen/Qwen2.5-7B-Instruct) — not loading again.
look.cmd_capture(model, tok, LAYER, NOTE) will use that copy.
NOTE: Patient received FOLFOX. Disease progressed. FOLFOX was discontinued.
LAYER: 20 ready


## 6. Run


look.cmd_capture(model, tok, LAYER, NOTE)


Yes — **this is now doing what you wanted.**

The two important pieces are:

> `Qwen already loaded on :8270 pid=762992 ... — not loading again.`

and then your actual experiment successfully ran:

> `look.cmd_capture(model, tok, LAYER, NOTE)`

with:

```text
prompt tokens: 15
shape: (1, 15, 3584)
layer 20 last-token L2: 114.6233
first 10: [...]
```

So you have demonstrated the critical thing:

```text
               :8270
         ONE LIVE QWEN
          PID 762992
               │
               ▼
        Jupyter notebook
               │
               ▼
      look.cmd_capture(...)
               │
               ▼
     actual L20 activation
       3584 dimensions
       L2 = 114.6233
```

And notice something even better: **114.6233 is the same L20 residual magnitude you were getting in the earlier experiment.** That's a strong sanity check that you're accessing the expected model/experiment path rather than accidentally analyzing some unrelated object.

So **do not change this working path right now.**

Your next step for the GUI is now much simpler. Cursor should take this exact proven mechanism and put the controls around it:

```text
Clinical Note
[ Patient received FOLFOX... ]

Layer
[ 20 ]

Experiment
[ Capture ▼ ]

[ RUN ]

          ↓

look.cmd_capture(...)
using :8270 Qwen

          ↓

Results displayed here
```

Then add the other experiments one at a time:

**Capture → Sites → A/B → layer sweep → MLP-zero → patch → steering → SAE/circuit experiments.**

You have crossed the important architectural hurdle: **a notebook experiment is now running while reporting that it uses the Qwen already owned by `:8270`, rather than loading another copy.** Now the GUI should be built *around this working execution path*, not redesigned again.


## 7–8. What happened / What did I learn?

_(fill after you run)_

## 9. Claim boundary

✓ We ran an Observe measurement.

✗ Observe ≠ causal proof. No intervene yet in this notebook.

## 10. CXR connection

`look.py --mode capture`

## 11. Questions

## 12. Revision notes

| Date | Change |
|------|--------|
| | |
